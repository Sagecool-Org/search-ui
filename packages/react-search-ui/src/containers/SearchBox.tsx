import React, { useState } from "react";
import {
  SearchBox,
  SearchBoxContainerProps
} from "@elastic/react-search-ui-views";
import { useSearch } from "../hooks";

const SearchBoxContainer = ({
  autocompleteMinimumCharacters = 0,
  autocompleteResults,
  autocompleteSuggestions,
  className,
  autocompleteView,
  inputProps,
  inputView,
  onSelectAutocomplete,
  shouldClearFilters = true,
  onSubmit,
  searchAsYouType,
  debounceLength,
  view,
  ...rest
}: SearchBoxContainerProps) => {
  const {
    autocompletedResults,
    autocompletedSuggestions,
    searchTerm,
    setSearchTerm,
    trackAutocompleteClickThrough,
    trackAutocompleteSuggestionClickThrough
  } = useSearch();

  const [isFocused, setIsFocused] = useState<boolean>(false);
  const handleFocus = () => {
    setIsFocused(true);
  };
  const handleBlur = () => {
    setIsFocused(false);
  };
  const handleSubmit = (e) => {
    e.preventDefault();
    setSearchTerm(searchTerm, {
      shouldClearFilters
    });
  };

  const handleChange = (value) => {
    const options = {
      autocompleteMinimumCharacters,
      ...((autocompleteResults ||
        autocompleteSuggestions ||
        searchAsYouType) && {
        debounce: debounceLength || 200
      }),
      shouldClearFilters,
      refresh: !!searchAsYouType,
      autocompleteResults: !!autocompleteResults,
      autocompleteSuggestions: !!autocompleteSuggestions
    };

    setSearchTerm(value, options);
  };

  const handleNotifyAutocompleteSelected = (selection) => {
    // results
    if (autocompleteResults) {
      const autocompleteResultsConfig =
        autocompleteResults === true
          ? { clickThroughTags: [], shouldTrackClickThrough: true }
          : autocompleteResults;

      if (
        !selection.suggestion &&
        autocompleteResultsConfig.shouldTrackClickThrough !== false
      ) {
        const { clickThroughTags = [] } = autocompleteResultsConfig;
        const id = selection.id?.raw;
        trackAutocompleteClickThrough(id, clickThroughTags);
      }

      if (selection.suggestion) {
        trackAutocompleteSuggestionClickThrough(
          selection.suggestion,
          selection.index,
          []
        );
      }
    }
  };
  const completeSuggestion = (searchTerm) => {
    setSearchTerm(searchTerm, {
      shouldClearFilters
    });
  };
  const defaultOnSelectAutocomplete = (selection) => {
    if (!selection) return;

    handleNotifyAutocompleteSelected(selection);
    if (!selection.suggestion && typeof autocompleteResults !== "boolean") {
      const url = selection[autocompleteResults.urlField]
        ? selection[autocompleteResults.urlField].raw
        : "";
      if (url) {
        const target =
          (typeof autocompleteResults !== "boolean" &&
            autocompleteResults.linkTarget) ||
          "_self";
        window.open(url, target);
      }
    } else {
      completeSuggestion(selection.suggestion);
    }
  };
  const View = view || SearchBox;

  const useAutocomplete =
    (!!autocompleteResults || !!autocompleteSuggestions) &&
    searchTerm.length >= autocompleteMinimumCharacters;
  const autocompletedSuggestionsCount = Object.entries(
    autocompletedSuggestions
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
  ).reduce((acc, [_, value]: [any, any]) => acc + value.length, 0);

  const allAutocompletedItemsCount =
    autocompletedSuggestionsCount + autocompletedResults.length;
  let handleOnSelectAutocomplete;
  if (onSelectAutocomplete) {
    handleOnSelectAutocomplete = (selection) => {
      onSelectAutocomplete(
        selection,
        {
          notifyAutocompleteSelected: handleNotifyAutocompleteSelected,
          completeSuggestion: completeSuggestion,
          autocompleteResults: autocompleteResults
        },
        defaultOnSelectAutocomplete
      );
    };
  }
  const viewProps = {
    allAutocompletedItemsCount: allAutocompletedItemsCount,
    autocompleteView,
    autocompleteResults: autocompleteResults,
    autocompleteSuggestions: autocompleteSuggestions,
    autocompletedResults: autocompletedResults,
    autocompletedSuggestions: autocompletedSuggestions,
    className,
    autocompletedSuggestionsCount: autocompletedSuggestionsCount,
    completeSuggestion: completeSuggestion,
    isFocused: isFocused,
    notifyAutocompleteSelected: handleNotifyAutocompleteSelected,
    onChange: (value) => handleChange(value),
    onSelectAutocomplete:
      handleOnSelectAutocomplete || defaultOnSelectAutocomplete,
    onSubmit: onSubmit
      ? (e) => {
          e.preventDefault();
          onSubmit(searchTerm);
        }
      : handleSubmit,
    useAutocomplete: useAutocomplete,
    value: searchTerm,
    inputProps: {
      onFocus: handleFocus,
      onBlur: handleBlur,
      ...inputProps
    },
    inputView,
    ...rest
  };
  return <View {...viewProps} />;
};
export default SearchBoxContainer;
