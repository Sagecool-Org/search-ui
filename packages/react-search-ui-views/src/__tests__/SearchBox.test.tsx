import React from "react";
import { render } from "@testing-library/react";
import SearchBox from "../SearchBox";
import { SearchBoxViewProps } from "../SearchBox";

const requiredProps: SearchBoxViewProps = {
  completeSuggestion: () => ({}),
  onChange: () => ({}),
  onSubmit: () => ({}),
  allAutocompletedItemsCount: 0,
  autocompletedResults: [],
  autocompletedSuggestions: {},
  autocompletedSuggestionsCount: 0,
  notifyAutocompleteSelected: () => ({}),
  value: "test",
  inputProps: {
    onFocus: () => {},
    onBlur: () => {}
  },
  isFocused: false,
  onSelectAutocomplete: jest.fn(),
  useAutocomplete: true
};

it("renders correctly", () => {
  const { container } = render(<SearchBox {...requiredProps} />);
  expect(container).toMatchSnapshot();
});

it("renders correctly with single autocompleteSuggestion configuration", () => {
  const { container } = render(
    <SearchBox
      {...requiredProps}
      autocompleteSuggestions={{ sectionTitle: "test" }}
    />
  );
  expect(container).toMatchSnapshot();
});

it("applies 'focused' class when `isFocused` is true", () => {
  const { container } = render(
    <SearchBox {...requiredProps} isFocused={true} />
  );
  const input = container.querySelector(".sui-search-box__text-input");
  expect(input).toHaveClass("focus");
});

it("passes through downshiftProps", () => {
  const { container } = render(
    <SearchBox
      {...requiredProps}
      inputProps={{
        placeholder: "test",
        onBlur: jest.fn(),
        onFocus: jest.fn()
      }}
    />
  );
  const input = container.querySelector(".sui-search-box__text-input");
  expect(input).toBeInTheDocument();
});

it("passes through inputProps", () => {
  const { container } = render(
    <SearchBox {...requiredProps} inputProps={{ placeholder: "test" }} />
  );
  const input = container.querySelector(".sui-search-box__text-input");
  expect(input).toHaveAttribute("placeholder", "test");
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const { container } = render(
    <SearchBox {...requiredProps} className={customClassName} />
  );
  expect(container.querySelector(".sui-search-box")).toHaveClass(
    "sui-search-box",
    "test-class"
  );
});

it("applies className from inputProps to input element", () => {
  const customClassName = "test-class";
  const { container } = render(
    <SearchBox {...requiredProps} inputProps={{ className: customClassName }} />
  );
  const input = container.querySelector(".sui-search-box__text-input");
  expect(input).toHaveClass("sui-search-box__text-input", "test-class");
});

describe("inputView prop", () => {
  let container, input, button;

  function setup() {
    const result = render(
      <SearchBox
        {...requiredProps}
        inputView={({ getAutocomplete, getInputProps, getButtonProps }) => {
          return (
            <>
              <div className="some_custom_wrapper_class">
                <input
                  {...getInputProps({
                    className: "some_custom_input_class"
                  })}
                />
                {getAutocomplete()}
                <input
                  {...getButtonProps({
                    className: "some_custom_button_class"
                  })}
                />
              </div>
            </>
          );
        }}
      />
    );
    container = result.container;
    input = container.querySelector("input.some_custom_input_class");
    button = container.querySelector("input.some_custom_button_class");
  }

  it("will render a custom view just for the input section", () => {
    setup();
    expect(
      container.querySelector(".some_custom_wrapper_class")
    ).toBeInTheDocument();
  });

  describe("when getInputProps is used", () => {
    it("will render custom props on input", () => {
      setup();
      expect(input).toHaveClass("some_custom_input_class");
    });

    it("will not overwrite the base class on input", () => {
      setup();
      expect(input).toHaveClass("sui-search-box__text-input");
    });

    it("will render base props on input", () => {
      setup();
      expect(input).toHaveAttribute("placeholder", "Search");
    });
  });

  describe("when getButtonProps is used", () => {
    it("will render custom props on button", () => {
      setup();
      expect(button).toHaveClass("some_custom_button_class");
    });

    it("will not overwrite the base class on button", () => {
      setup();
      expect(button).toHaveClass("sui-search-box__submit");
    });

    it("will render base props on button", () => {
      setup();
      expect(button).toHaveAttribute("type", "submit");
    });
  });
});
