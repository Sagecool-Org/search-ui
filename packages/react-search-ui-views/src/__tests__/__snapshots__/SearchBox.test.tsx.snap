// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <form
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    role="combobox"
  >
    <div
      class="sui-search-box"
    >
      <div
        class="sui-search-box__wrapper"
      >
        <input
          aria-autocomplete="list"
          aria-labelledby="downshift-0-label"
          autocomplete="off"
          class="sui-search-box__text-input "
          data-transaction-name="search input"
          id="downshift-0-input"
          placeholder="Search"
          value="test"
        />
      </div>
      <input
        class="button sui-search-box__submit"
        data-transaction-name="search submit"
        type="submit"
        value="Search"
      />
    </div>
  </form>
</div>
`;

exports[`renders correctly with single autocompleteSuggestion configuration 1`] = `
<div>
  <form
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-1-label"
    role="combobox"
  >
    <div
      class="sui-search-box"
    >
      <div
        class="sui-search-box__wrapper"
      >
        <input
          aria-autocomplete="list"
          aria-labelledby="downshift-1-label"
          autocomplete="off"
          class="sui-search-box__text-input "
          data-transaction-name="search input"
          id="downshift-1-input"
          placeholder="Search"
          value="test"
        />
      </div>
      <input
        class="button sui-search-box__submit"
        data-transaction-name="search submit"
        type="submit"
        value="Search"
      />
    </div>
  </form>
</div>
`;
